Issue while linking Python 3.7 libraries when `-lintl` from `gettext` is linked

In the autoconf build we rely on `sysconfig.get_config_var` to find out the python link libraries:\. With Python 3.6 we get:

```bash
$ python3.6 -c "import sysconfig; print (sysconfig.get_config_var('LIBS'))"
-ldl -lm -lpthread -lutil
```

On our cluster with Python 3.7 we get:

```
 $ python3.7 -c "import sysconfig; print (sysconfig.get_config_var('LIBS'))"
-lcrypt -lintl -lpthread -ldl  -lutil
```

Note that extra `libint` library. This comes from gettext library which Python uses for [internationalization](https://docs.python.org/3/library/gettext.html). Because of this, when we build NEURON on our BB5 cluster, we get:

```bash
$ ../configure --with-nrnpython=`which python3` --disable-rx3d
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /usr/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether to enable maintainer-specific portions of Makefiles... yes
checking for style of include used by make... GNU

get_config_var('LIBDIR')  '/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/python-3.7.4-tfxecymrkn/lib'
get_config_var('LIBS')  '-lcrypt -lintl -lpthread -ldl  -lutil'
get_config_var('MODLIBS')  '' returning ''
checking if python include files and libraries work... configure: error: could not run a test that used the python library.
Examine config.log to see error details. Something wrong with
	PYLIB=-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/python-3.7.4-tfxecymrkn/lib -lpython3.7m -lcrypt -lintl -lpthread -ldl  -lutil  -R/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/python-3.7.4-tfxecymrkn/lib
or
	PYLIBDIR=/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/python-3.7.4-tfxecymrkn/lib
or
	PYLIBLINK=-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/python-3.7.4-tfxecymrkn/lib -lpython3.7m -lcrypt -lintl -lpthread -ldl  -lutil
or
	PYINCDIR=/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/python-3.7.4-tfxecymrkn/include/python3.7m
```

So why `libintl` is not found? If we look at how Python is built, gettext is passed as LDFLAGS :

```
==> [2020-02-01-05:55:53.209836] '/nvme/bbprelman/84306/spack-stage-python-3.7.4-tfxecymrknrdwnem536oajs5wzjjznv4/spack-src/configure'
'LDFLAGS=-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/expat-2.2.9-rkc5cz3vef/lib'
'--enable-shared' 
'--without-ensurepip' 
'CFLAGS=-fPIC' '--with-openssl=/usr' 
'--with-dbmliborder=gdbm:bdb:ndbm' 
'--with-system-expat' 
'--with-system-ffi'
```

So if we want to use `sysconfig.get_config_var('LIBS')` directly, we need to provide path of `gettext`. We can find all necessary `LDFLAGS` using `sysconfig.get_config_var` again:

```
$ python3.7 -c "import sysconfig; print (sysconfig.get_config_var('LDFLAGS'))"
-L/usr/lib64 
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/tools/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/gdbm-1.14.1-frv7kcgfty/lib 
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/expat-2.2.9-rkc5cz3vef/lib 
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/readline-8.0-uuhvgs2hgl/lib 
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/ncurses-6.1-esk7yfugjx/lib 
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/tools/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/gettext-0.19.8.1-ngmcb6sv3g/lib 
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/zlib-1.2.11-5jk2xjgi67/lib 
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/xz-5.2.4-yl7j3yvx5w/lib 
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/sqlite-3.30.1-2sdl4kjxw2/lib 
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/bzip2-1.0.8-urnhhbekzr/lib 
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/libffi-3.2.1-gbrqfaxhzz/lib64
```
So in [this commit](https://github.com/neuronsimulator/nrn/pull/453/commits/df336f1858e37e8266c1789b71e20411eeb5619a) I attempted to add `LDFLAGS` to Python library link line. Configure works fine and I thought the issue is solved. But while building I am seeing:

```
/bin/sh ../../libtool  --tag=CXX   --mode=link
/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/compilers/2020-02-01/linux-rhel7-x86_64/gcc-4.8.5/gcc-8.3.0-r24zbv3waz/bin/g++
-pthread -g -O2   -o nrniv nvkludge.o ../ivoc/nrnmain.o ../ivoc/ivocmain.o
../oc/modlreg.o ../oc/ockludge.o ../nrnoc/libnrnoc.la ../oc/liboc.la libnrniv.la
../ivoc/libivoc.la ../oc/liboc.la ../nrnmpi/libnrnmpi.la
../gnu/libneuron_gnu.la ../scopmath/libscopmath.la ../sparse13/libsparse13.la
../sundials/libsundials.la ../memacs/libmemacs.la ../mesch/libmeschach.la
../ivos/libivos.la -lX11  ../nrnpython/libnrnpython.la -L/usr/lib64
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/tools/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/gdbm-1.14.1-frv7kcgfty/lib
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/expat-2.2.9-rkc5cz3vef/lib
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/readline-8.0-uuhvgs2hgl/lib
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/ncurses-6.1-esk7yfugjx/lib
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/tools/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/gettext-0.19.8.1-ngmcb6sv3g/lib
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/zlib-1.2.11-5jk2xjgi67/lib
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/xz-5.2.4-yl7j3yvx5w/lib
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/sqlite-3.30.1-2sdl4kjxw2/lib
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/bzip2-1.0.8-urnhhbekzr/lib
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/libffi-3.2.1-gbrqfaxhzz/lib64
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/python-3.7.4-tfxecymrkn/lib
-lpython3.7m -lcrypt -lintl -lpthread -ldl  -lutil
-R/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/python-3.7.4-tfxecymrkn/lib
-lreadline -lncurses    -lm -ldl libtool: link:
/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/compilers/2020-02-01/linux-rhel7-x86_64/gcc-4.8.5/gcc-8.3.0-r24zbv3waz/bin/g++
-pthread -g -O2 -o .libs/nrniv nvkludge.o ../ivoc/nrnmain.o ../ivoc/ivocmain.o
../oc/modlreg.o ../oc/ockludge.o  ../nrnoc/.libs/libnrnoc.so ./.libs/libnrniv.so
../ivoc/.libs/libivoc.so ../oc/.libs/liboc.so ../nrnmpi/.libs/libnrnmpi.so
../gnu/.libs/libneuron_gnu.so ../scopmath/.libs/libscopmath.so
../sparse13/.libs/libsparse13.so ../sundials/.libs/libsundials.so
../memacs/.libs/libmemacs.so ../mesch/.libs/libmeschach.so
../ivos/.libs/libivos.so -lX11 ../nrnpython/.libs/libnrnpython.so
/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/compilers/2020-02-01/linux-rhel7-x86_64/gcc-4.8.5/gcc-8.3.0-r24zbv3waz/lib/../lib64/libstdc++.so
-L/usr/lib64
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/tools/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/gdbm-1.14.1-frv7kcgfty/lib
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/expat-2.2.9-rkc5cz3vef/lib
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/readline-8.0-uuhvgs2hgl/lib
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/ncurses-6.1-esk7yfugjx/lib
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/tools/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/gettext-0.19.8.1-ngmcb6sv3g/lib
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/zlib-1.2.11-5jk2xjgi67/lib
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/xz-5.2.4-yl7j3yvx5w/lib
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/sqlite-3.30.1-2sdl4kjxw2/lib
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/bzip2-1.0.8-urnhhbekzr/lib
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/libffi-3.2.1-gbrqfaxhzz/lib64
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/python-3.7.4-tfxecymrkn/lib
-lpython3.7m -lcrypt
/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/tools/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/gettext-0.19.8.1-ngmcb6sv3g/lib/libintl.so
-L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/tools/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/libxml2-2.9.9-imarll7dfh/lib
-liconv -lc -lpthread -lutil -lreadline -lncurses -lm -ldl -pthread -Wl,-rpath
-Wl,/usr/local/nrn/x86_64/lib -Wl,-rpath
-Wl,/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/compilers/2020-02-01/linux-rhel7-x86_64/gcc-4.8.5/gcc-8.3.0-r24zbv3waz/lib/../lib64
-Wl,-rpath
-Wl,/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/tools/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/gettext-0.19.8.1-ngmcb6sv3g/lib
-Wl,-rpath
-Wl,/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/python-3.7.4-tfxecymrkn/lib
/usr/bin/ld: cannot find -liconv
```

So `where does -liconv comes from`? The problem is that when we pass `libintl` to libtool it find `.la` file which has:

```
$ cat ... gettext-0.19.8.1-ngmcb6sv3g/lib/libintl.la
...
# Names of this library.
library_names='libintl.so.8.1.5 libintl.so.8 libintl.so'

# The name of the static archive.
old_library='libintl.a'

# Linker flags that cannot go in dependency_libs.
inherited_linker_flags=''

# Libraries that this one depends upon.
dependency_libs=' -L/gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/tools/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/libxml2-2.9.9-imarll7dfh/lib 
-liconv -lc'

# Names of additional weak libraries provided by this library
```

So libtool inserts dependency from `.la` and this goes on and on. (I didn't look further).

In case of CMake everything works fine because [FindPythonLibs](https://github.com/neuronsimulator/nrn/blob/master/cmake/modules/FindPythonLibsNew.cmake) module we have just look for Python library : 

```
-- Found PythonLibs: /gpfs/bbp.cscs.ch/ssd/apps/hpc/jenkins/deploy/external-libraries/2020-02-01/linux-rhel7-x86_64/gcc-8.3.0/python-3.7.4-tfxecymrkn/lib/libpython3.7m.so
```

So what are possible solutions?

* autoconf build should not use all libraries returned by`LIBS`? at least `-lintl` can be safely removed. I did that in [next commit ](https://github.com/neuronsimulator/nrn/pull/453/commits/e69598fadb5a9208779334096e1141f138a82884).
* user can explicitly specific LDFLAGS with library path of gettext. But if we could make it work automatically it would better as I can see this issue appearing on other clusters as well.

--- a/.travis.yml
+++ b/.travis.yml
@@ -100,7 +100,7 @@ install:
       export CYTHON_LIB=$($PIP show cython | grep Location);
       export CYTHON_EXE=$(echo $CYTHON_LIB | awk '{print $2}' | awk -F "lib" '{print$1}')"bin";
       export PATH=$CYTHON_EXE:$PATH;
-      echo $PATH;
+      brew link --overwrite python;
     fi
 
 #=============================================================================

--- a/m4/nrnpython.m4
+++ b/m4/nrnpython.m4
@@ -42,7 +42,7 @@ Examine config.log to see error details. Something wrong with
 	PYLIB=$PYLIB
 or
 	PYLIBDIR=$PYLIBDIR
-or	
+or
 	PYLIBLINK=$PYLIBLINK
 or
 	PYINCDIR=$PYINCDIR
@@ -212,6 +212,12 @@ explicitly specify PYINCDIR])
 				PYLIBDIR="$gcfLIBDIR"
 				PYLIBLINK="-L$PYLIBDIR -l$PYLIB $gcfLIBS $gcfMODLIBS"
 				PYLIB="$PYLIBLINK -R$PYLIBDIR"
+			fi
+			dnl add LDFLAGS for python version >= 3.7
+			if [[ $(bc <<< "$PYVER >= 3.7") -eq 1 ]] ; then
+				AC_NRN_PYCONF(EXTRALDFLAGS, get_config_var('LDFLAGS'),"",$ac_nrn_python)
+				PYLIBLINK="$EXTRALDFLAGS $PYLIBLINK"
+				PYLIB="$EXTRALDFLAGS $PYLIB"
 			fi
 		fi
 

--- a/m4/nrnpython.m4
+++ b/m4/nrnpython.m4
@@ -192,7 +192,7 @@ explicitly specify PYINCDIR])
 			fi
 		fi
 		if test "$EXTRAPYLIBS" = "" ; then
-			AC_NRN_PYCONF(EXTRAPYLIBS,get_config_var('LIBS'),"",$ac_nrn_python)
+			AC_NRN_PYCONF(EXTRAPYLIBS,get_config_var('LIBS').replace('-lintl',''),"",$ac_nrn_python)
 		fi
 		setup_extra_link_args=extra_link_args
 		case "$host_os" in
@@ -205,19 +205,13 @@ explicitly specify PYINCDIR])
 		if test "$PYLIB" = "" ; then
 			AC_NRN_PYCONF(gcfLIBRARY, get_config_var('LIBRARY'),"",$ac_nrn_python)
 			AC_NRN_PYCONF(gcfLIBDIR, get_config_var('LIBDIR'),"",$ac_nrn_python)
-			AC_NRN_PYCONF(gcfLIBS, get_config_var('LIBS'),"",$ac_nrn_python)
+			AC_NRN_PYCONF(gcfLIBS, get_config_var('LIBS').replace('-lintl',''),"",$ac_nrn_python)
 			AC_NRN_PYCONF(gcfMODLIBS, get_config_var('MODLIBS'),"",$ac_nrn_python)
 			PYLIB=`echo $gcfLIBRARY|sed 's/lib\(.*\)\.a/\1/'`
 			if test "$PYLIB" != "" ; then
 				PYLIBDIR="$gcfLIBDIR"
 				PYLIBLINK="-L$PYLIBDIR -l$PYLIB $gcfLIBS $gcfMODLIBS"
 				PYLIB="$PYLIBLINK -R$PYLIBDIR"
-			fi
-			dnl add LDFLAGS for python version >= 3.7
-			if [[ $(bc <<< "$PYVER >= 3.7") -eq 1 ]] ; then
-				AC_NRN_PYCONF(EXTRALDFLAGS, get_config_var('LDFLAGS'),"",$ac_nrn_python)
-				PYLIBLINK="$EXTRALDFLAGS $PYLIBLINK"
-				PYLIB="$EXTRALDFLAGS $PYLIB"
 			fi
 		fi
 

